Add H.265 packetizer and related codec selection

ZitaLiao opened the pull request.

murillo128 commented:
This PR adds H265 packetization according to RFC 7798 RTP Payload Format Implementation. Our plan is to integrate in OBS so it can be used with the nvenc h265 encoder which provides awesome quality.

We are testing/validating the PR with the fixed H265  support Safari as per https://github.com/WebKit/WebKit/pull/15494

@paullouisageneau could you review this and provide us feedback of any change that may be required for merging?

paullouisageneau commented on include/rtc/h265packetizationhandler.hpp:
```diff
@@ -0,0 +1,32 @@
+namespace rtc {
+
+/// Handler for H265 packetization
+class RTC_CPP_EXPORT H265PacketizationHandler final : public MediaChainableHandler {
```
I'm currently refactoring media handlers to simplify the design in https://github.com/paullouisageneau/libdatachannel/pull/929. One of the main changes will be that packetization handlers are not necessary anymore.

This could be merged before (in which case I'll refactor it too) or after, I'm fine with both.

ZitaLiao replied:
Hi @paullouisageneau , thanks for your review :) I just rebase today, and didn't merge https://github.com/paullouisageneau/libdatachannel/pull/929 since it's still open.

paullouisageneau commented on include/rtc/h265rtppacketizer.hpp:
```diff
@@ -0,0 +1,61 @@
+	inline static const uint32_t defaultClockRate = 90 * 1000;
+
+	/// NAL unit separator
+	enum class Separator {
```
This should be shared with `H264RtpPacketizer`.

ZitaLiao replied:
Put it into class NalUnit to fix this comment.
A better refactor would make NalUnit a virtual class then make H264 and H265 implement it. But to make safer and incremental change, I just leave H264 as it is now, making only H265NalUnit inherent from the NalUnit class.

paullouisageneau commented on src/h265rtppacketizer.cpp:
```diff
@@ -0,0 +1,167 @@
+	NUSM_longMatch
+} NalUnitStartSequenceMatch;
+
+NalUnitStartSequenceMatch StartSequenceMatchSucc(NalUnitStartSequenceMatch match, byte _byte,
```
This should be shared with `H264RtpPacketizer`, for instance a common implementation in `RtpPacketizer`.

ZitaLiao replied:
Fixed.

paullouisageneau reviewed (commented):
It looks good, thank you for the contribution! I've just a few comments about duplicated code which could be made common with the h264 packetizer.

ZitaLiao commented:
@paullouisageneau Thank you for your review!
I rebased today and fixed per your comment. Could you please help to review again please? Thank you!

ZitaLiao commented:
@paullouisageneau Thank you for the review! 
I rebased today and fixed per your comment. Could you please help to review again? Thank you in advanced.

paullouisageneau reviewed (commented):
Thank you for the changes, it looks good overall. Don't worry about https://github.com/paullouisageneau/libdatachannel/pull/929, I'll add this to the refactoring.

ZitaLiao commented on include/rtc/h264rtppacketizer.hpp:
```diff
@@ -27,15 +27,7 @@ class RTC_CPP_EXPORT H264RtpPacketizer final : public RtpPacketizer,
 	inline static const uint32_t defaultClockRate = 90 * 1000;
 
-	/// NAL unit separator
-	enum class Separator {
```
Fixed. Please check.

ZitaLiao commented on include/rtc/h265nalunit.hpp:
```diff
@@ -0,0 +1,181 @@
+/**
+ * Copyright (c) 2020 Filip Klembara (in2core)
```
Fix. Please check

paullouisageneau reviewed (approved):
Thanks!
